src/typegraph/graph/src/conv/mod.rs

```diff
@@ -0,0 +1,375 @@
+        let reg = std::mem::take(ne.registry());
+        Ok(crate::Typegraph {
+            schema,
+            root: root.assert_object()?.clone(),
+            input_types: conv.input_types,
+            output_types: conv.output_types,
+            functions: conv.functions,
+            namespace_objects: conv.namespace_objects,
+            named: reg.map,
+            conversion_map: conv
+                .conversion_map
+                .direct
+                .into_iter()
+                .map(TryFrom::try_from)
+                .collect::<Result<Vec<_>>>()?,
+            runtimes: conv.runtimes,
+            materializers: conv.materializers,
+        })
+    }
+
+    fn register_converted_type(
+        &mut self,
+        rpath: RelativePath,
+        ty: Type,
+        duplicate: bool, // -> recursive
+    ) -> Result<()> {
+        if !duplicate {
+            self.conversion_map.register(rpath.clone(), ty.clone())?;
+            // recursivity will be handled by the finalizer
+        } else if !matches!(ty, Type::Function(_)) {
+            // value type
+
+            let added = self.conversion_map.append(ty.key(), rpath.clone())?;
+            if added {
+                let cyclic = rpath.pop().contains(&self.schema.types, &From::from(&ty));
+                if !cyclic {
+                    for edge in ty.edges() {
+                        let child_ty = edge.to;
+                        if let Ok(seg) = edge.kind.try_into() {
+                            self.register_converted_type(rpath.push(seg)?, child_ty, true)?;
+                        }
+                    }
+                }
+            }
+        }
+
+        use RelativePath as RP;
+        match rpath {
+            RP::Function(fn_idx) => {
+                self.functions.insert(fn_idx, ty.assert_func()?.clone());
+            }
+            RP::NsObject(path) => {
+                self.namespace_objects
+                    .insert(path.clone(), ty.assert_object()?.clone());
+            }
+            RP::Input(_) => {
+                self.input_types.insert(ty.key(), ty.clone());
+            }
+            RP::Output(_) => {
+                self.output_types.insert(ty.key(), ty.clone());
+            }
+        }
+
+        Ok(())
+    }
+
+    fn register_type_2<C: FnOnce(TypeKey) -> Box<dyn TypeConversionResult>>(
+        &mut self,
+        rpath: RelativePath,
+        key: TypeKey,
+        convert: C,
+        duplicate: bool,
+    ) -> Result<Box<dyn TypeConversionResult>> {
+        let conv_res = convert(key);
+        let typ = conv_res.get_type();
+
+        self.register_converted_type(rpath.clone(), typ.clone(), duplicate)?;
+
+        Ok(conv_res)
+    }
+
+    fn register_type<C: FnOnce(TypeKey) -> Box<dyn TypeConversionResult>>(
+        &mut self,
+        rpath: RelativePath,
+        key: TypeKey,
+        convert: C,
+    ) -> Result<Box<dyn TypeConversionResult>> {
+        self.register_type_2(rpath, key, convert, false)
+    }
```

Comment:
suggestion: Add more context maybe like renaming to `register_unique_type` or similar

Reply:
This was just a temporary function.. I'll fix it.